src/crawlers/browser_crawler.js

```diff
@@ -0,0 +1,523 @@
+            // @TODO: Maybe we could also automatically set persistCookiesPerSession to false when useSessionPool is false and log warning
+            throw new Error('You cannot use "persistCookiesPerSession" without "useSessionPool" set to true.');
+        }
+
+        super({
+            ...basicCrawlerOptions,
+            useSessionPool: false,
+            handleRequestFunction: (...args) => this._handleRequestFunction(...args),
+            handleRequestTimeoutSecs: handlePageTimeoutSecs * BASIC_CRAWLER_TIMEOUT_MULTIPLIER,
+        });
+
+        this.handlePageFunction = handlePageFunction;
+        this.handlePageTimeoutSecs = handlePageTimeoutSecs;
+        this.handlePageTimeoutMillis = this.handlePageTimeoutSecs * 1000;
+
+        this.gotoFunction = gotoFunction;
+        this.gotoOptions = {};
+
+        this.persistCookiesPerSession = persistCookiesPerSession;
+        this.proxyConfiguration = proxyConfiguration;
+
+        this.preNavigationHooks = preNavigationHooks;
+        this.postNavigationHooks = postNavigationHooks;
+
+        if (useSessionPool) {
+            this.sessionPool = new SessionPool({
+                ...sessionPoolOptions,
+                log: this.log,
+            });
+
+            // Assuming there are not more thant 20 browsers running at once;
+            this.sessionPool.setMaxListeners(20);
+        }
+
+        const { preLaunchHooks = [], postLaunchHooks = [], ...rest } = browserPoolOptions;
+        this.browserPool = new BrowserPool({
+            ...rest,
+            preLaunchHooks: [
+                this._extendLaunchContext.bind(this),
+                ...preLaunchHooks,
+            ],
+            postLaunchHooks: [
+                this._maybeAddSessionRetiredListener.bind(this),
+                ...postLaunchHooks,
+            ],
+        });
+    }
+
+    /**
+     * Wrapper around handlePageFunction that opens and closes pages etc.
+     *
+     * @param {Object} crawlingContext
+     * @param {Request} crawlingContext.request
+     * @param {BrowserController} crawlingContext.browserController
+     * @param {Session} [crawlingContext.session]
+     * @private
+     */
+    async _handleRequestFunction(crawlingContext) {
+        const { id } = crawlingContext;
+        const page = await this.browserPool.newPage({ id });
+        this._enhanceCrawlingContextWithPageInfo(crawlingContext, page);
+
+        const { request, session } = crawlingContext;
+
+        if (this.persistCookiesPerSession) {
+            const cookies = crawlingContext.session.getPuppeteerCookies(request.url);
+            await crawlingContext.browserController.setCookies(page, cookies);
+        }
+
+        try {
+            await this._handleNavigation(crawlingContext);
+
+            await this._responseHandler(crawlingContext);
+
+            // save cookies
+            // @TODO: Should we save the cookies also after/only the handle page?
+            if (this.persistCookiesPerSession) {
```

Comment:
I have no idea what the comments mean

Reply:
It means that you can change the cookies in the handle page function and those are not saved automatically. I can imagine that user behavioral tracking could be updating cookies this way. Right now these cookies would be ignored by default.